fast_isio:
  # topics
  sonarPointCloudTopic: "/BlueRov2/0/sonar_edge_feature_laserscan"
  sonarImageTopic: "/BlueRov2/0/sonar"
  imuTopic: "/BlueRov2/0/imu"
  odomTopic: "/odometry/imu"
  poseGtTopic: "/BlueRov2/0/position"

  # Frames
  sonarFrame: "sonar_link"
  baselinkFrame: "base_link"
  odometryFrame: "odom"
  mapFrame: "map"

  # GPS Settings
  useImuHeadingInitialization: false           # if using GPS data, set to "true"
  useGpsElevation: false                      # if GPS elevation is bad, set to "false"
  gpsCovThreshold: 2.0                        # m^2, threshold for using GPS data
  poseCovThreshold: 25.0                      # m^2, threshold for using GPS data
  addGpsInterval: 2                           # m, interval for adding GPS data 


  # IMU Settings
  imuAccNoise: 7.5406250915386773e-04
  imuGyrNoise: 1.4086669484740333e-02
  imuAccBiasN: 2.5388117684307040e-05
  imuGyrBiasN: 1.3738859212852894e-03
  imuGravity: 0
  imuRPYWeight: 0.01

  # Extrinsics (Sonar -> IMU)
  extrinsicTrans: [-0.1, 0.0, 0.0]
  extrinsicRPY: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]
  extrinsicRot: [1, 0, 0,
                  0, 1, 0,
                  0, 0, 1]

  # Sonar Settings
  N_SCAN: 128                                 # number of sonar channel       (128)
  Horizon_SCAN: 512                           # sonar horizontal resolution    (512)
  downsampleRate: 1                           # default: 1. Downsample your data if too many points. i.e., 16 = 64 / 4, 16 = 16 / 1
  sonarMinRange: 0.1                          # default: 1.0, minimum lidar range to be used
  sonarMaxRange: 15

  # voxel filter paprams
  odometrySurfLeafSize: 0.05                   
  mappingCornerLeafSize: 0.025  
  mappingSurfLeafSize: 0.05       

  # sonar feature threshold
  edgeThreshold: 0.05
  surfThreshold: 0.05
  edgeFeatureMinValidNum: 10
  surfFeatureMinValidNum: 100    

  # Surrounding map
  surroundingkeyframeAddingDistThreshold:  0.5  # meters, regulate keyframe adding threshold   (0.5)
  surroundingkeyframeAddingAngleThreshold: 0.2  # radians, regulate keyframe adding threshold  (0.2)
  surroundingKeyframeDensity: 1.0               # meters, downsample surrounding keyframe poses   
  surroundingKeyframeSearchRadius: 30.0         # meters, within n meters scan-to-map optimization (when loop closure disabled)     

  numberOfCores: 4                              # number of cores for mapping optimization
  mappingProcessInterval: 0.1                   # seconds, regulate mapping frequency    

  z_tollerance: 1000                               # meters (0)
  rotation_tollerance: 1000                        # radians (1)


  # Loop closure
  loopClosureEnableFlag: true
  loopClosureFrequency: 100                     # Hz, regulate loop closure constraint add frequency
  surroundingKeyframeSize: 50                   # submap size (when loop closure enabled)
  historyKeyframeSearchRadius: 50               # meters, key frame that is within n meters from current pose will be considerd for loop closure
  historyKeyframeSearchTimeDiff: 50             # seconds, key frame that is n seconds older will be considered for loop closure
  historyKeyframeSearchNum: 50                  # number of hostory key frames will be fused into a submap for loop closure
  historyKeyframeFitnessScore: 0.3              # icp threshold, the smaller the better alignment

  # Visualization
  globalMapVisualizationSearchRadius: 1000.0    # meters, global map visualization radius
  globalMapVisualizationPoseDensity: 1          # meters, global map visualization keyframe density
  globalMapVisualizationLeafSize: 0.1           # meters, global map visualization cloud density